File: examples/legacy/perl/gmrpl.pl
 
# Copyright 2007-2021 United States Government as represented by the
# Administrator of The National Aeronautics and Space Administration.
# No copyright is claimed in the United States under Title 17, U.S. Code.
# All Rights Reserved.







 
# gmrpl.pl
#
#
# A Perl example demonstratioin of GMSEC reply functionality.



use lib '../../bin/lib';
use strict;
use GMSECAPI;
use Util;


{
   my $task = bless({});
   Util::driver($task, @ARGV);
}


sub printUsage
{
   my ($self, $config) = @_;
   my $hasConnectionType = Util::hasConnectionType($config);

   return 0 if ($hasConnectionType);

   print "\n\nusage: perl gmrpl.pl connectiontype=<middleware> ";
   print "[ <parameter>=<value> ]";
   print "\n\n\tNote that the parameter 'connectiontype' is required.";
   print "\n\tThe rest of other parameters are optional.";
   print "\n\n\tserver=<server name>";
   print " (required if middleware is not bolt/MB locally)";
   print "\n\tsubject=<subject name>";
   print "\n\tmsg_timeout_ms=<timeout period (milliseconds)>";
   print "\n\tprog_timeout_s=<timeout period (seconds)>";
   print "\n\t\tNote: msg_timeout_ms must be specified for prog_timeout_s";
   print "\n\tcfgfile=<config_filename.xml>";
   print "\n\nFor more information, see API User's Guide\n\n";

   return 1;
}


sub run
{
   my ($self, $config) = @_;

   my $subject = Util::getString($config, 'SUBJECT', 'GMSEC.>'); 
   my $msg_timeout_ms = Util::getInteger($config, 'MSG_TIMEOUT_MS', 
                                         $GMSEC_WAIT_FOREVER);
                                         
   my $prog_timeout_s = Util::getInteger($config, 'PROG_TIMEOUT_S',
   										 $GMSEC_WAIT_FOREVER);
   
   if($msg_timeout_ms == $GMSEC_WAIT_FOREVER && 
						$prog_timeout_s != $GMSEC_WAIT_FOREVER)
   {
		Util::report("Usage: To use prog_timeout_s, msg_timeout_ms must be specified!");
		return 1;
   }   

   # Output GMSEC API version
   Util::report(GMSECAPI::ConnectionFactory::GetAPIVersion());

   # Create the Connection
   my $connection = GMSECAPI::ConnectionFactory::Create($config);
   Util::check('ConnectionFactory.Create',
       GMSECAPI::ConnectionFactory::Status());
   $self->{CONNECTION} = $connection;

   # Connect
   unless ($connection->Connect) {
     Util::check('Connect', $connection->Status);
   }


   # output connection middleware version
   Util::report("Middleware version = " . $connection->GetLibraryVersion());

   # Subscribe
   Util::report("Subscribing to " . $subject);
   unless ($connection->Subscribe($subject)) {
     Util::check('Subscribe', $connection->Status);
   }

   Util::report("Subscribing to " . "GMSEC.TERMINATE" . "\n");
   unless ($connection->Subscribe('GMSEC.TERMINATE')) {
     Util::check('Subscription to GMSEC.TERMINATE failed', $connection->Status);
   }

   my $done = 0;
   my $prevTime = time();
   my $nextTime;
   my $elapsedTime = 0;

   # Listen
   while (not $done)
   {
   
	  if($elapsedTime >= $prog_timeout_s && $prog_timeout_s != $GMSEC_WAIT_FOREVER)
	  {
	     $done = 1;
	     next;
	  }
	  
      my $message = $connection->GetNextMsg($msg_timeout_ms);
      my $status = $connection->Status;
	  
	  if($prog_timeout_s != $GMSEC_WAIT_FOREVER)
	  {
	  	 $nextTime = time();
	     $elapsedTime += ($nextTime - $prevTime);
	     $prevTime = $nextTime;
	  }
	  
      $self->{MESSAGE} = $message;

      if ($status->isError())
      {
         if ($status->GetCode() == $GMSEC_TIMEOUT_OCCURRED)
         {
            Util::checkNoThrow('GetNextMsg(' . $msg_timeout_ms . ')', $status);
            next;
         }
         Util::checkThrow('GetNextMsg(' . $msg_timeout_ms . ')', $status);
      }

      my $temp_subject = $message->GetSubject;
      Util::check('GetSubject()', $message->Status);
      $done = 1 if ($temp_subject eq "GMSEC.TERMINATE");

      # Dump the message
      print "received =>\n", $message->ToXML, "\n";

      # check for request
      my $kind = $message->GetKind;

      if ($GMSEC_MSG_REQUEST == $kind ) 
      {
         my $temp_subject = $message->GetSubject;

         # Construct Reply subject.
         my $messageField = $message->GetField("COMPONENT");
         my $compVal = $messageField->GetValue();

         $messageField = $message->GetField("MSG-ID");
         my $idVal = $messageField->GetValue();

         # Set Status Code.  
         # See API Interface Specification for available status codes.
         my $statusCode = 1;

         my $reply_subject = "GMSEC.MISSION.SAT_ID.RESP.".$compVal.
                             ".".$idVal.".".$statusCode;

         # Create Reply if not done
         my $reply;
         if (not $done)
         {
            $reply = $connection->CreateMessage($reply_subject, 
                                                $GMSEC_MSG_REPLY);
         }
         else
         {
            $reply = $connection->CreateMessage($temp_subject, 
                                                $GMSEC_MSG_REPLY);
         }
         Util::check('CreateMessage(reply)', $connection->Status);
         $self->{REPLY} = $reply;

         # Add Fields
         my $field = GMSECAPI::Field->new($GMSEC_TYPE_STRING, "ANSWER",
                                          "Sure looks like it!");
         unless ($reply->AddField($field)) {
            Util::check('AddField(ANSWER)', $reply->Status);
         }

         $self->{FIELD} = $field;

         # Add the name of the component sending the Reply.
         $field->SetName("COMPONENT");
         $field->SetValue("GMRPL");
         unless ($reply->AddField($field)) {
            Util::check('AddField(COMPONENT)', $reply->Status);
         }

         # Dump Reply
         print "prepared reply =>\n", $reply->ToXML, "\n";

         # Send Reply
         unless ($connection->Reply($message, $reply)) {
            Util::check('Connection->Reply', $connection->Status);
         }

         print "\nsent reply =>\n";

         my $temp_msg = $message;
         $message = $NULL;
         $self->{MESSAGE} = $message;

         # Destroy inbound message
         unless ( $connection->DestroyMessage($temp_msg) ) {
            Util::check('DestroyMessage', $connection->Status);
         }
      }
   }

}


sub cleanup
{
   my ($self) = @_;

   my $reply = $self->{REPLY};
   my $message = $self->{MESSAGE};
   my $connection = $self->{CONNECTION};

   # Destroy the message
   if ($message && $message->isValid())
   {
      print "destroying message\n";
      unless ($connection->DestroyMessage($message)) {
         Util::check('DestroyMessage', $connection->Status);
      }
   }

   # Destroy the reply message
   if ($reply && $reply->isValid())
   {
      print "destroying reply message\n";
      unless ($connection->DestroyMessage($reply)) {
         Util::check('DestroyMessage', $connection->Status);
      }
   }

   if ($connection)
   {
      # Disconnect if necessary
      if ($connection->IsConnected)
      {
         print "disconnecting\n";
         unless ($connection->Disconnect) {
            Util::check('Disconnect', $connection->Status);
         }
      }

      # Destroy the Connection
      print "destroying connection\n";
      unless (GMSECAPI::ConnectionFactory::Destroy($connection)) {
         Util::check('ConnectionFactory.Destroy',
                     GMSECAPI::ConnectionFactory::Status());
      }
   }
}


1;

__END__

File: examples/legacy/java/gmrpl.java
 
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */






 
/** @file gmrpl.java
 *
 * An example replier using the Java GMSEC API.
 */

import gov.nasa.gsfc.gmsecapi.Config;
import gov.nasa.gsfc.gmsecapi.Connection;
import gov.nasa.gsfc.gmsecapi.Field;
import gov.nasa.gsfc.gmsecapi.GMSEC_UShort;
import gov.nasa.gsfc.gmsecapi.GMSEC_String;
import gov.nasa.gsfc.gmsecapi.Message;
import gov.nasa.gsfc.gmsecapi.Status;
import static gov.nasa.gsfc.gmsecapi.gmsecAPI.*;
import gov.nasa.gsfc.gmsecapi.util.*;


public class gmrpl implements Example
{
	static final String DEFAULT_SUBJECT = "GMSEC.>";

	StatusChecker checker = new StatusChecker();
	Config config;
	Connection connection;
	Message message;
	Message reply;


	gmrpl(Config config) throws ExampleException {
		this.config = config;
		Util.initialize(config);

		if (!Util.hasConnectionType(config)) {
			printUsage();
			throw new ExampleException("Invalid option");
		}
	}

	private void printUsage() {
		Log.Error("\nusage: java gmrpl connectionType=<middleware> " +
				"[ <parameter>=<value> ]\n" +
				"\n\tNote that the parameter 'connectiontype' is required. "  +
				"\n\tThe rest of other parameters are optional.\n" +
				"\n\tserver=<server name> " +
				"(required if middleware is not bolt/MB locally)" +
				"\n\tsubject=<subject name>" +
				"\n\tmsg_timeout_ms=<timeout period (milliseconds)>" +
				"\n\tprog_timeout_s=<timeout period (seconds)>" +
				"\n\t\tNote: msg_timeout_ms must be specified for prog_timeout_s" +
				"\n\tloglevel=<log level>" +
				"\n\tcfgfile=<config_filepath.xml>" +
				"\n\nFor more information, see API User's Guide\n\n");
	}

	public boolean run() throws ExampleException {
		Status status;

		int msg_timeout_ms = Util.get(config,  "MSG_TIMEOUT_MS", 
				Util.EX_MSG_TIMEOUT_MS);
		int prog_timeout_s = Util.get(config, "PROG_TIMEOUT_S",
				Util.EX_PROG_TIMEOUT_S);
        if(msg_timeout_ms == GMSEC_WAIT_FOREVER && 
									prog_timeout_s != GMSEC_WAIT_FOREVER)
        {
           Log.Error("Usage: To use prog_timeout_s, msg_timeout_ms must be defined!");
           return false;
        }  	

		connection = Util.openConnection(config);

		String subject = Util.get(config, "SUBJECT", DEFAULT_SUBJECT);
		Log.Info("Subscribing to " + subject);
		checker.calling("Subscribe(" + subject + ")")
			.check(connection.Subscribe(subject));

		Log.Info("Subscribing to GMSEC.TERMINATE");
		checker.calling("Subscribe(" + "GMSEC.TERMINATE" + ")")
			.check(connection.Subscribe("GMSEC.TERMINATE"));

		boolean done = false;
		long prevTime = System.currentTimeMillis();
		long nextTime = 0;
		int elapsedTime = 0;	//ms

		while (!done) {

			if(elapsedTime >= (prog_timeout_s * 1000)
								 && prog_timeout_s != GMSEC_WAIT_FOREVER)
			{
				done = true;
				Log.Info("Program Timeout Reached");
				continue;
			}
			
			message = new Message();

			checker.calling("GetNextMsg(timeout=" + msg_timeout_ms + ")");
			status = connection.GetNextMsg(message, msg_timeout_ms);
			
			nextTime = System.currentTimeMillis();
			elapsedTime += (int) (nextTime - prevTime);
			prevTime = nextTime;
			
			if (status.isError()) {
				if (status.GetCode() == GMSEC_TIMEOUT_OCCURRED)
					Log.Info("timeout occurred");
				else
					Util.check("GetNextMsg", status);
			}
			else {
				GMSEC_String tempSubject = new GMSEC_String();
				checker.calling("GetSubject").check(message.GetSubject(tempSubject));
				if (tempSubject.Get().equals("GMSEC.TERMINATE"))
					done = true;

				GMSEC_String s = new GMSEC_String();
				checker.calling("Message.ToXML").check(message.ToXML(s));

				Log.Info("received\n" + s);

				GMSEC_UShort kind = new GMSEC_UShort();
				checker.calling("Message.GetKind").check(message.GetKind(kind));
				if (GMSEC_MSG_REQUEST == kind.Get()) {

					String component = getStringField(message, "COMPONENT");
					if (component == null)
						throw new ExampleException("request missing COMPONENT");

					String id = getStringField(message, "MSG-ID");
					if (id == null)
						throw new ExampleException("request missing MSG-ID");

					// Set Status Code.  See API Interface Specification for available status codes.
					String replyStatus = "1";
					String replySubject = "GMSEC.MISSION.SAT_ID.RESP." + component + "." + id + "." + replyStatus;

					// Create Reply
					reply = new Message();
					checker.calling("CreateMessage:reply")
						.check(connection.CreateMessage(replySubject, GMSEC_MSG_REPLY, reply));

					// Add fields to reply
					Field field = new Field();
					field.SetFieldValueSTRING("ANSWER", "Sure looks like it!");
					checker.calling("AddField:ANSWER").check(reply.AddField(field));

					// Add the name of the component sending the Reply.
					field.SetFieldValueSTRING("COMPONENT", "GMRPL");
					checker.calling("AddField:COMPONENT").check(reply.AddField(field));

					//..dump msg
					GMSEC_String replyXML = new GMSEC_String();
					checker.calling("reply.ToXML").check(reply.ToXML(replyXML));
					Log.Info("prepared reply =>\n" + replyXML.Get());

					// Send Reply
					checker.calling("Connection.Reply")
							.check(connection.Reply(message, reply));
					Log.Info("reply sent!");

					Message tmp = message;
					message = null;
					checker.calling("DestroyMessage:received").check(connection.DestroyMessage(tmp));

					tmp = reply;
					reply = null;
					checker.calling("DestroyMessage:reply").check(connection.DestroyMessage(tmp));
				}
			}
		}

		return true;
	}
		

	public boolean cleanup() throws ExampleException {
		if (message != null && message.isValid()) {
			Util.checkNoThrow("DestroyMessage", connection.DestroyMessage(message));
		}
		
		if (reply != null && reply.isValid()) {
			Util.checkNoThrow("DestroyMessage", connection.DestroyMessage(reply));
		}
		
		if (connection != null)
			Util.closeConnection(connection);

		return false;
	}

	private String getStringField(Message message, String key) {
		String value = null;
		Field field = new Field();
		Status status = message.GetField(key, field);
		if (Util.checkNoThrow("GetField:" + key, status)) {
			GMSEC_String tmp = new GMSEC_String();
			status = field.GetValueSTRING(tmp);
			if (Util.checkNoThrow("Field.GetStringValue:" + key, status))
				value = tmp.Get();
		}
		return value;
	}

	public static void main(String[] args) {

		Example example = null;

		try {
			Config config = new Config(args);

			Util.addToConfigFromFile(config);

			example = new gmrpl(config);
			example.run();
		}
		catch (ExampleException e) {
			e.printStackTrace();
		}
		finally {
			if (example != null)
				Util.cleanup(example);
		}

	}

}


